src/combinatorics/catalan.lean

```diff
@@ -0,0 +1,158 @@
+/-
+Copyright (c) 2022 Julian Kuelshammer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Julian Kuelshammer
+-/
+import data.nat.choose.central
+import algebra.big_operators.fin
+import tactic.field_simp
+
+/-!
+# Catalan numbers
+
+The Catalan numbers (http://oeis.org/A000108) are probably the most ubiquitous sequence of integers
+in mathematics. They enumerate several important objects like binary trees, Dyck paths, and
+triangulations of convex polygons.
+
+## Main definitions
+
+* `catalan n`: the `n`th Catalan number, defined recursively as
+               `catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i)`.
+* `catalan' n`: the `n`th Catalan number, defined explicitly as
+               `catalan' n = 1 / (n + 1) * nat.central_binom n`.
+
+## Main results
+
+* `catalan_defs_agree`: The fact that the two definitions both give the same sequence.
+
+## Implementation details
+
+The proof of the equality of the two definitions follows
+https://math.stackexchange.com/questions/3304415/catalan-numbers-algebraic-proof-of-the-recurrence-relation
+
+## TODO
+
+* Prove that the Catalan numbers enumerate many interesting objects.
+* Generalise to Catalan numbers associated to arbitrary Coxeter groups.
+
+-/
+
+open_locale big_operators
+open finset
+
+/-- The recursive definition of the sequence of Catalan numbers:
+`catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i)` -/
+def catalan : ℕ → ℕ
+| 0       := 1
+| (n + 1) := ∑ i : fin n.succ, have _ := i.2, have _ := nat.lt_succ_iff.mpr (n.sub_le i),
+             catalan i * catalan (n - i)
+
+@[simp] lemma catalan_zero : catalan 0 = 1 := by rw catalan
+
+lemma catalan_succ (n : ℕ) : catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i) :=
+by rw catalan
+
+/-- The explicit definition of the sequence of Catalan numbers via the central binomial
+coefficient.-/
+def catalan' (n : ℕ) :  ℚ :=
+1 / (n + 1) * nat.central_binom n
+
+@[simp] lemma catalan'_zero : catalan' 0 = 1 :=
+begin rw catalan', simp, end
+
+/-- A helper sequence that can be used to prove the equality of the recursive and the explicit
+definition using a telescoping sum argument. -/
+private def gosper_catalan (n j : ℕ) : ℚ :=
```

Comment:
What does this sequence represent? Is it just an auxiliary construction that is useless outside of these proofs, or could it potentially be used elsewhere? In the latter case, it should be unprivated, made into a sequence of naturals, and given a more descriptive name.

Reply:
To my knowledge this is just an auxiliary construction, it appears if you apply https://en.wikipedia.org/wiki/Gosper%27s_algorithm to the sequence `catalan i * catalan (n-i)`.